### PR TITLE
Add byte-optimized implementations for reading and writing int and long values

### PR DESCRIPTION
**Pull request by boaks**
Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramReader.java [hunk 1]
```diff
@@ -180,26 +180,34 @@ public void close() {
...
+			throw new IllegalArgumentException("bits must be in range 0 ... 64!");
+		}
 		long bits = 0; // initialize all bits to zero
 
-		for (int i = numBits - 1; i >= 0; i--) {
-
-			// check whether new byte needs to be read
-			if (currentBitIndex < 0) {
-				readCurrentByte();
+		if (currentBitIndex < 0 && (numBits & 0x7) == 0) {
```
Comment (sbernard31) [inline element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramReader.java]: `(numBits & 0x7) == 0`, I think the idea is to ensure that numbits is a multiple of 8.
For me using modulo is more intuitive.

(but probably subjective)

File: element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramReader.java [hunk 2]
```diff
@@ -180,26 +180,34 @@ public void close() {
...
+			for (int i = numBits; i > 0; i -= 8) {
```
Comment (sbernard31) [inline element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramReader.java]: Why not, the classical way ?
``` java
for (int i = 0; i < numBits; i+= 8) {
}
```

File: element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramReader.java [hunk 3]
```diff
@@ -180,26 +180,34 @@ public void close() {
...
+				bits <<= 8;
+				bits |= readByte();
```
Comment (sbernard31) [inline element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramReader.java]: Should we add a comment just to say  this if block is optimization ?

**Comment by boaks** [comment]
> Did you observe any benefits of this ?

I hoped so, but the benchmark fails to show it ;-).

**Comment by boaks** [comment]
`// byte-wise, no maverick bits left`

is that comment ok? other proposal?

**Comment by sbernard31** [comment]
:+1:

**Pull request merged**